python/cugraph/structure/new_number_map.py

```diff
@@ -0,0 +1,275 @@
+import cugraph.comms as Comms
+import dask_cudf
+import numpy as np
+import cudf
+
+
+def call_renumber(sID,
+                  data,
+                  num_verts,
+                  num_edges,
+                  is_mnmg):
+    wid = Comms.get_worker_id(sID)
+    handle = Comms.get_handle(sID)
+    return c_renumber.renumber(data[0],
+                               num_verts,
+                               num_edges,
+                               wid,
+                               handle,
+                               is_mnmg)
+
+class NumberMap:
+
+    class SingleGPU:
+        def __init__(self, df, src_col_names, dst_col_names, id_type,
+                     store_transposed):
+            self.col_names = NumberMap.compute_vals(src_col_names)
+            self.df = cudf.DataFrame()
+            self.id_type = id_type
+            self.store_transposed = store_transposed
+            self.numbered = False
+
+        def to_internal_vertex_id(self, df, col_names):
+            tmp_df = df[col_names].rename(
+                columns=dict(zip(col_names, self.col_names)), copy=False
+            )
+            index_name = NumberMap.generate_unused_column_name(df.columns)
+            tmp_df[index_name] = tmp_df.index
+            return (
+                self.df.merge(tmp_df, on=self.col_names, how="right")
+                .sort_values(index_name)
+                .drop(columns=[index_name])
+                .reset_index()["id"]
+            )
+
+        def from_internal_vertex_id(
+            self, df, internal_column_name, external_column_names
+        ):
+            tmp_df = self.df.merge(
+                df,
+                right_on=internal_column_name,
+                left_on="id",
+                how="right",
+            )
+            if internal_column_name != "id":
+                tmp_df = tmp_df.drop(columns=["id"])
+            if external_column_names is None:
+                return tmp_df
+            else:
+                return tmp_df.rename(
+                    columns=dict(zip(self.col_names, external_column_names)),
+                    copy=False,
+                )
+
+    class MultiGPU:
+        def __init__(
+            self, ddf, src_col_names, dst_col_names, id_type, store_transposed
+        ):
+            self.col_names = NumberMap.compute_vals(src_col_names)
+            self.val_types = NumberMap.compute_vals_types(ddf, src_col_names)
+            self.val_types["count"] = np.int32
+            self.id_type = id_type
+            self.store_transposed = store_transposed
+            self.numbered = False
+
+        def to_internal_vertex_id(self, ddf, col_names):
+            return self.ddf.merge(
+                ddf,
+                right_on=col_names,
+                left_on=self.col_names,
+                how="right",
+            )["global_id"]
+
+        def from_internal_vertex_id(
+            self, df, internal_column_name, external_column_names
+        ):
```

Comment:
I think this can be much faster. External vertex id = local_map[internal vertex id - base vertex ID for this GPU];

Reply:
Will be addressed as a part of optimization in subsequent PRs